File: A2Z_striver/patterns/pattern_3/pattern_3.py
import sys


class Solution:
    """
    Problem: Print a Right-Angled Triangle of increasing numbers.
    Example Input: n = 4
    Example Output (Space separated):
    1
    1 2
    1 2 3
    1 2 3 4
    
    Constraints: 1 <= n <= 100
    Goal: O(N^2) Time.
    """

    # // ~ Approach 1: Nested Loops (Universal Logic)
    # // Analysis: Time O(N^2) | Space O(1)
    def pattern3_nested(self, n):
        for i in range(n):
            # // . Inner loop runs i+1 times
            for j in range(i + 1):
                # // ! Logic: Print (j + 1) because j starts at 0
                # // - end=" " ensures readability for multi-digit numbers
                print(j + 1, end=" ")
            
            # // - Move to next line
            print()

    # // ~ Approach 2: Pythonic One-Liner (String Join)
    # // Analysis: Time O(N^2) | Space O(N) (creates a temporary list per row)
    def pattern3_pythonic(self, n):
        for i in range(n):
            # // - Create a range 1 to i+1, convert to str, join with spaces
            row_str = " ".join(str(x) for x in range(1, i + 2))
            print(row_str)

# Driver Code
if __name__ == "__main__":
    solver = Solution()
    N = 5
    
    print(f"--- Approach 1: Nested Loops (N={N}) ---")
    solver.pattern3_nested(N)
    
    print(f"\n--- Approach 2: Pythonic (N={N}) ---")
    solver.pattern3_pythonic(N)

File: A2Z_striver/1 Basics/1.2 Patterns/pattern_3/pattern_3.cpp
#include <iostream>

using namespace std;

// Header Comment Block
// Problem: Print a Right-Angled Triangle of increasing numbers.
// Example Input: n = 4
// Example Output:
// 1
// 1 2
// 1 2 3
// 1 2 3 4
// Goal: O(N^2) Time | O(1) Space

class Solution
{
public:
    // ~ Approach: Dependent Nested Loops
    // Analysis: Time O(N^2) | Space O(1)
    void pattern3(int n)
    {
        for (int i = 0; i < n; i++)
        {
            // - Inner loop runs i+1 times (0 to i)
            for (int j = 0; j <= i; j++)
            {
                // ! Logic: Output the Column index (j) + 1
                // - Added " " for readability when n >= 10
                cout << j + 1 << " ";
            }
            cout << "\n";
        }
    }
};

int main()
{
    // ! Fast I/O
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;
    int n = 5;

    cout << "Test Case (n=" << n << "):\n";
    sol.pattern3(n);

    return 0;
}